Add "Quick Action Link" to the Dashboard in the Post List

jblz opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Include a link to the statistics page for an article in the wp-admin Posts List. This uses the [Find in Parse.ly API Endpoint](https://www.parse.ly/help/integration/other-tools#find-in-parsely-api-endpoint).

Our filtering function is hooked into both the `post_row_actions` & the `page_row_actions` so it will do its thing on hierarchical and non lists.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

To make it easier to browse to the article page in the dashboard from within the WordPress interface.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

* Browse to `/wp-admin/edit.php`
* There should be links to the Parse.ly Dashboard for each published post
* Browse to `/wp-admin/edit.php?post_type=page`
* There should be links to the Parse.ly Dashboard for each published page
* Browse to the wp-admin list for a "publicly-queryable" custom post type
* There should be links to the Parse.ly Dashboard for each published page
* Browse to the wp-admin list for a custom post type that is NOT "publicly-queryable"
* There should **NOT** be links to the Parse.ly Dashboard for each published page

## Screenshots (if appropriate):

|Link to `/find` endpoint|No Link on Private Posts|No Link on Drafts|
|---|---|---|
|![Screen Shot 2021-05-21 at 4 27 10 PM](https://user-images.githubusercontent.com/1587282/119198356-2f18af00-ba57-11eb-9ec2-80b35a962b95.png)|<img width="248" alt="Screen Shot 2021-05-24 at 2 22 42 PM" src="https://user-images.githubusercontent.com/1587282/119391121-f74e7900-bc9b-11eb-82c0-756b1209c7a6.png">|<img width="249" alt="Screen Shot 2021-05-24 at 2 22 49 PM" src="https://user-images.githubusercontent.com/1587282/119391145-fddcf080-bc9b-11eb-915b-409f951a4587.png">|




## Types of changes

New feature (non-breaking change which adds functionality)

hbbtstar committed:
add prototype find in parsely link

jblz committed:
whitespace

jblz committed:
Bail if post is not published or otherwise untrackable

jblz committed:
whitespace++

jblz committed:
wrap a string for translation

jblz committed:
yet another whitespace change

GaryJones commented on src/class-parsely.php:
```diff
@@ -92,6 +93,24 @@ public function __construct() {
 	}
 
+	function add_parsely_link( $actions, $post ) {
+		if ( ! self::post_has_trackable_status( $post ) ) {
+			return $actions;
+		}
```
While a post may not currently have a trackable status, it may have had one previously - and so there may be data in Parse.ly dashboard about it.

Is this an edge case that we should ignore for now?

jblz replied:
Yes, I think this is an ok edge case to ignore for now. If it becomes obvious how we can know if an article has been crawled, we can revisit.  I think we're going to want to do so if / when we make this indication more dynamic.

GaryJones commented on src/class-parsely.php:
```diff
@@ -92,6 +93,24 @@ public function __construct() {
 		add_action( 'wp_enqueue_scripts', [ $this, 'wp_parsely_style_init' ] );
 	}
 
+	function add_parsely_link( $actions, $post ) {
```
```suggestion
	public function add_parsely_link( $actions, $post ) {
```

Also needs a DocBlock.

GaryJones commented on src/class-parsely.php:
```diff
@@ -92,6 +93,24 @@ public function __construct() {
+		$options = $this->get_options();
+		$apikey = $options['apikey'];
+		$post_url = get_permalink( $post );
+		$parsely_url = 'https://dash.parsely.com/' . $apikey . '/find?url=' . $post_url;
```
The `$post_url` is going to need encoding.

GaryJones commented on src/class-parsely.php:
```diff
@@ -92,6 +93,24 @@ public function __construct() {
+		$parsely_url = 'https://dash.parsely.com/' . $apikey . '/find?url=' . $post_url;
+		$actions = array_merge( $actions, array(
+			'find_in_parsely' => sprintf(
+				'<img style="padding-right: 4px; transform: translateY(2px);" width="16px" height="16px" src="https://www.parse.ly/favicon-32x32.png?v=f82f85b2b9b3e85c26fc141c6c44b638"/><a href="%1$s">%2$s</a>',
```
```suggestion
				'<img style="padding-right: 4px; transform: translateY(2px);" width="16px" height="16px" src="https://www.parse.ly/favicon-32x32.png?v=f82f85b2b9b3e85c26fc141c6c44b638" alt="" /><a href="%1$s">%2$s</a>',
```

jblz committed:
Merge branch 'develop' of github.com:Parsely/wp-parsely into xand-add-quick-action-link

jblz committed:
Run filtering function on page_row_actions as well

jblz committed:
specify function name to row_actions_add_parsely_link

jblz committed:
add docblock & specify function is public visibility

jblz committed:
Link "front page" to main dash, urlencode query arg, rm img, esc link label

jblz committed:
tweak the comment

jblz commented:
I noticed a suboptimal experience when browsing to one of these links while not logged into the dashboard.

Logged here: https://github.com/Parsely/dash/issues/1038

jblz committed:
Don't check page_for_posts -- that should still link to the find page

GaryJones commented on src/class-parsely.php:
```diff
@@ -129,6 +131,52 @@ public function __construct() {
+		if ( ! is_array( $actions ) ) {
+			return $actions;
+		}
+
```
I think we should also check for the post type being `public` as well - if it's not accessible from the front-end, then it can't be in Parse.ly. A post could be non-public post type and still have a `publish` post status (WPCOM Legacy Redirector may go that route, to distinguish active from non-active redirects).

jblz replied:
Good suggestion.

It would be nice to just use [`is_post_publicly_viewable`](https://developer.wordpress.org/reference/functions/is_post_publicly_viewable/), but it was added in WP 5.7, so too new for our minimum.

Even [`is_post_type_viewable`](https://developer.wordpress.org/reference/functions/is_post_type_viewable/) requires WP >= 4.4.

What do you think we should check here?

GaryJones replied:
We could always either create our own function/check, or bump the minimum to WP 4.4 (or maybe later if we're going to move it anyway).

jblz replied:
5e4885659b9870a9379e40eb8a1f5e54f8ee6406

GaryJones commented on src/class-parsely.php:
```diff
@@ -129,6 +131,52 @@ public function __construct() {
+	 *
+	 * @see https://developer.wordpress.org/reference/hooks/page_row_actions/
+	 * @see https://developer.wordpress.org/reference/hooks/post_row_actions/
+	 * @param [string[]] $actions  The list of actions to which this function will append
+	 * @param [WP_Post]  $post     The individual post object in the list
```
```suggestion
	 * @param array<string, string> $actions The existing list of actions.
	 * @param WP_Post               $post    The individual post object the actions apply to.
```

`string[]` is only for indexed array, like
`[ 'foo', 'bar' ]`. 

What we have in actions, is an associative array, with a string key and a string value: 
`[ 'foo' => 'a link', 'parsely_dash' => 'another link' ]`

GaryJones commented on src/class-parsely.php:
```diff
@@ -129,6 +131,52 @@ public function __construct() {
+	 * @see https://developer.wordpress.org/reference/hooks/post_row_actions/
+	 * @param [string[]] $actions  The list of actions to which this function will append
+	 * @param [WP_Post]  $post     The individual post object in the list
+	 * @return $actions            The filtered list of actions
```
```suggestion
	 * @return array<string, string> The amended list of actions.
```

We return a value, not a variable.

GaryJones commented on src/class-parsely.php:
```diff
@@ -129,6 +131,52 @@ public function __construct() {
+		}
+
+		$options = $this->get_options();
+		$apikey = $options['apikey'];
+		$parsely_url = trailingslashit( 'https://dash.parsely.com/' . $apikey );
```
Should we check that apikey has been set before assuming it has? Can't create a link if we don't know where we're meant to link to.

GaryJones commented on src/class-parsely.php:
```diff
@@ -129,6 +131,52 @@ public function __construct() {
+		}
+
+		$post_url = get_permalink( $post );
+		$parsely_url .=  'find?url=' . urlencode( $post_url );
```
Does Parse.ly require `urlencode()`'d querystring, or `rawurlencode()`'d querystring? They do give slightly different results.

jblz replied:
I think we may want to switch to [`rawurlencode`](https://www.php.net/manual/en/function.rawurlencode.php) for the full [RFC3986](http://www.faqs.org/rfcs/rfc3986.html) treatment.  See also:
https://stackoverflow.com/a/996161

@hbbtstar -- do you have any insight into what the service expects?

jblz replied:
I went ahead and switched to `rawurlencode`. I think that's probably preferable, but we can change it back if needed.

GaryJones commented on src/class-parsely.php:
```diff
@@ -129,6 +131,52 @@ public function __construct() {
+			'page' === get_option( 'show_on_front' ) &&
+			(int) get_option( 'page_on_front' ) === $post->ID
+		) {
+			// This is the "front page." Link to the dashboard.
+			$actions[ 'parsely_dash' ] = sprintf(
+				'<a href="%1$s">%2$s</a>',
+				esc_url( $parsely_url ),
+				esc_html__( 'Parse.ly Dashboard', 'wp-parsely' )
+			);
```
Why does this need to be different?

The front page still has it's own stats page on the Parse.ly dashboard i.e. `dash.parsely.com/wpvip.com/find?url=https://wpvip.com` still works.

Having a list of pages pointing to "Parse.ly Stats", and then one in the middle pointing to "Parse.ly Dashboard" seems odd. Let's keep it consistent.

jblz replied:
I went ahead and took out the variation.  Agree that having it consistent makes more sense after more testing.

jblz committed:
fix type in docblock

Co-authored-by: Gary Jones <gary.jones@automattic.com>

jblz committed:
Fix return type in docblock

Co-authored-by: Gary Jones <gary.jones@automattic.com>

jblz committed:
Don't add link if apikey is not set

jblz committed:
don't treat front page differently, build url w/ add_query_arg, use rawurlencode

jblz committed:
Ensure post type is public

jblz commented:
As of now, this includes the link on password-protected posts.

GaryJones commented on src/class-parsely.php:
```diff
@@ -129,6 +131,53 @@ public function __construct() {
+			trailingslashit( 'https://dash.parsely.com/' . $options['apikey'] ) . 'find'
+		);
+
+		$actions['find_in_parsely'] = sprintf(
+			'<a href="%1$s">%2$s</a>',
+			esc_url( $parsely_url ),
+			esc_html__( 'Parse.ly Stats', 'wp-parsely' )
+		);
```
To have the same link text, pointing to different URLs, is a bad thing for accessibility. Links can be taken out of context, so they need to be more explainable.

Let's look at what WP core does for the existing Edit, Quick Edit, Trash, and View actions links:

<img width="1626" alt="Screenshot 2021-05-26 at 11 23 01" src="https://user-images.githubusercontent.com/88371/119644608-f4c35f00-be14-11eb-8a0f-57a020298288.png">

They use an `aria-label` attribute on the anchor element, that includes the (quoted) name of the post in some way. We should do the same for this Parse.ly Stats link.

GaryJones replied:
Addressed in c6032d9.

GaryJones reviewed (changes requested):
We have an opportunity to maintain the level of accessibility that other row actions have.

GaryJones committed:
Row action: add aria-label to make link accessible

To have the same link text (such as "Parse.ly stats") pointing to different URLs, is a bad thing for accessibility. Links can be taken out of context, so they need to be more differentiated and unique.

We follow the same principle as what WordPress core does for the existing row action links, and that is to include an `aria-label` attribute which includes the title of the post. Visually, the link looks the same.

jblz committed:
Merge branch 'develop' into xand-add-quick-action-link
